File: fronted/app/src/main/java/com/asflum/cashewregister/google/auth/GoogleAuthHandler.kt
package com.asflum.cashewregister.google.auth

import android.content.Context
import androidx.credentials.Credential
import androidx.credentials.CredentialManager
import androidx.credentials.CredentialOption
import androidx.credentials.CustomCredential
import androidx.credentials.GetCredentialRequest
import com.asflum.cashewregister.BuildConfig
import com.google.android.libraries.identity.googleid.GetGoogleIdOption
import com.google.android.libraries.identity.googleid.GoogleIdTokenCredential

object GoogleAuthHandler {
    const val WEB_CLIENT_ID = BuildConfig.WEB_CLIENT_ID

    suspend fun getUserIdToken(context: Context): Result<String> {
        val credentialManager = CredentialManager.create(context)
        val request = GetCredentialRequest.Builder()
            .addCredentialOption(getCredentialOptions())
            .build()

        return try {
            val result = credentialManager.getCredential(context, request)
            val credential = result.credential

            val idToken = extractIdTokenFromCredential(credential)
            Result.success(idToken)
        } catch (e: Exception) {
            Result.failure(e)
        }
    }

    private fun getCredentialOptions(): CredentialOption {
        return GetGoogleIdOption.Builder()
            .setFilterByAuthorizedAccounts(false)
            .setAutoSelectEnabled(false)
            .setServerClientId(WEB_CLIENT_ID)
            .build()
    }

    private fun extractIdTokenFromCredential(credential: Credential): String {
        return when (credential) {
            is GoogleIdTokenCredential -> credential.idToken

            is CustomCredential -> {
                if (credential.type == GoogleIdTokenCredential.TYPE_GOOGLE_ID_TOKEN_CREDENTIAL) {
                    val googleCred = GoogleIdTokenCredential.createFrom(credential.data)
                    googleCred.idToken
                } else {
                    throw UnsupportedOperationException("Tipo de credencial personalizada no soportada")
                }
            }

            else -> throw IllegalArgumentException("Tipo de credencial desconocida")
        }
    }
}

File: frontend/frontend_flutter/expense_tracker/lib/services/google_auth_handler.dart
import 'package:credential_manager/credential_manager.dart';
import 'package:expense_tracker/utils/result.dart';
import 'package:flutter_dotenv/flutter_dotenv.dart';

class GoogleAuthHandler {
  static final CredentialManager _credentialManager = CredentialManager();

  static Future<Result<String>> getUserIdToken() async {
    try {
      if (_credentialManager.isSupportedPlatform) {
        await _credentialManager.init(
          preferImmediatelyAvailableCredentials: true,
          googleClientId: dotenv.get("WEB_CLIENT_ID"),
        );
      }

      final GoogleIdTokenCredential? credential = await _credentialManager
          .saveGoogleCredential(useButtonFlow: false);

      final idToken = credential?.idToken;

      if (idToken == null) {
        return Result.failure(Exception("No se obtuvo el ID Token."));
      }

      return Result.success(idToken);
    } on CredentialException catch (_) {
      return Result.failure(Exception("Error al guardar la credencial"));
    } catch (e) {
      return Result.failure(Exception("Error inesperado: $e"));
    }
  }
}
